From 54c113d2402495cf22d7b82f2c6eb14efa9bd1ee Mon Sep 17 00:00:00 2001
From: Liu ShuoX <shuox.liu@intel.com>
Date: Thu, 13 Jun 2013 13:50:09 +0800
Subject: [PATCH 072/429] intel_soc_pmu.c: correct the sub class of display on
 Merrifield

sub class of display pci device is different.
The one for MFLD and CLV is 0x00, and for Merrifield is 0x80.

Signed-off-by: Liu ShuoX <shuox.liu@intel.com>
Signed-off-by: Srinidhi Kasagar <srinidhi.kasagar@intel.com>
---
 arch/x86/platform/intel-mid/intel_soc_pmu.c | 21 +++++++++++++++++++--
 1 file changed, 19 insertions(+), 2 deletions(-)

diff --git a/arch/x86/platform/intel-mid/intel_soc_pmu.c b/arch/x86/platform/intel-mid/intel_soc_pmu.c
index c9c413c79263..b43ba7a29642 100644
--- a/arch/x86/platform/intel-mid/intel_soc_pmu.c
+++ b/arch/x86/platform/intel-mid/intel_soc_pmu.c
@@ -505,6 +505,21 @@ static inline u32 find_index_in_hash(struct pci_dev *pdev, int *found)
 	return h_index;
 }
 
+static bool is_display_subclass(unsigned int sub_class)
+{
+	/* On MDFLD and CLV, we have display PCI device class 0x30000,
+	 * On MRFLD, we have display PCI device class 0x38000
+	 */
+
+	if ((sub_class == 0x0 &&
+		(platform_is(INTEL_ATOM_MFLD) ||
+		platform_is(INTEL_ATOM_CLV))) ||
+		(sub_class == 0x80 && platform_is(INTEL_ATOM_MRFLD)))
+		return true;
+
+	return false;
+}
+
 static int get_pci_to_pmu_index(struct pci_dev *pdev)
 {
 	int pm, type;
@@ -533,7 +548,8 @@ static int get_pci_to_pmu_index(struct pci_dev *pdev)
 	type = ss & LOG_SS_MASK;
 	ss = ss & LOG_ID_MASK;
 
-	if ((base_class == PCI_BASE_CLASS_DISPLAY) && !sub_class)
+	if ((base_class == PCI_BASE_CLASS_DISPLAY) &&
+			is_display_subclass(sub_class))
 		index = 1;
 	else if ((base_class == PCI_BASE_CLASS_MULTIMEDIA) &&
 			(sub_class == ISP_SUB_CLASS))
@@ -597,7 +613,8 @@ static void get_pci_lss_info(struct pci_dev *pdev)
 		return;
 
 	/* initialize gfx subsystem info */
-	if ((base_class == PCI_BASE_CLASS_DISPLAY) && !sub_class) {
+	if ((base_class == PCI_BASE_CLASS_DISPLAY) &&
+			is_display_subclass(sub_class)) {
 		set_mid_pci_log_id(index, (u32)index);
 		set_mid_pci_cap(index, PM_SUPPORT);
 	} else if ((base_class == PCI_BASE_CLASS_MULTIMEDIA) &&
-- 
2.37.3

